File: flutter/shraed_prefskmp/lib/shared_prefs.dart
import 'package:shraed_prefskmp/shared_prefs_params.dart';

class SharedPrefs {
  Future<String?> getString(String key) async {
    final String? value = await PrefsConst.channel.invokeMethod(
      PrefsConst.getString,
      {PrefsParams.key: key},
    );
    return value;
  }

  Future<void> setString(String key, String value) async {
    await PrefsConst.channel.invokeMethod(
      PrefsConst.setString,
      {PrefsParams.key: key, PrefsParams.value: value},
    );
  }

  Future<int?> getInt(String key, {int defaultValue = 0}) async {
    final int? value = await PrefsConst.channel.invokeMethod(
      PrefsConst.getInt,
      {PrefsParams.key: key, PrefsParams.defaultt: defaultValue},
    );
    return value;
  }

  Future<void> setInt(String key, int value) async {
    await PrefsConst.channel.invokeMethod(
      PrefsConst.setInt,
      {PrefsParams.key: key, PrefsParams.value: value},
    );
  }

  Future<bool?> getBoolean(String key, {bool defaultValue = false}) async {
    final value = await PrefsConst.channel.invokeMethod(
      PrefsConst.getBoolean,
      {PrefsParams.key: key, PrefsParams.defaultt: defaultValue},
    );
    if (value == null) {
      return null;
    } else {
      if (value is bool) {
        return value;
      } else if (value is int) {
        if (value == 0) {
          return false;
        } else {
          return true;
        }
      }
    }
    return value;
  }

  Future<void> setBoolean(String key, bool value) async {
    await PrefsConst.channel.invokeMethod(
      PrefsConst.setBoolean,
      {PrefsParams.key: key, PrefsParams.value: value},
    );
  }

  Future<void> remove(String key) async {
    await PrefsConst.channel.invokeMethod(
      PrefsConst.remove,
      {PrefsParams.key: key},
    );
  }

  Future<void> clear() async {
    await PrefsConst.channel.invokeMethod(PrefsConst.clear);
  }

  Future<List<String>?> getStringList(String key) async {
    final List<Object?>? value = await PrefsConst.channel.invokeMethod(
      PrefsConst.getStringList,
      {PrefsParams.key: key},
    );
    return value?.cast<String>();
  }

  Future<void> setStringList(String key, List<String> values) async {
    await PrefsConst.channel.invokeMethod(
      PrefsConst.setStringList,
      {PrefsParams.key: key, PrefsParams.values: values},
    );
  }

  Future<bool> contains(String key) async {
    final bool exists = await PrefsConst.channel.invokeMethod(
      PrefsConst.contains,
      {PrefsParams.key: key},
    );
    return exists;
  }
}

File: kotlin/SharedPrefsKmp/shared/src/iosMain/kotlin/com/mohaberabi/sharedprefskmp/SharedPrefs.kt
package com.mohaberabi.sharedprefskmp


import platform.Foundation.NSUserDefaults

actual class SharedPrefs {


    companion object {
        private const val DOMAIN_NAME = "kmpPrefs"
    }

    private val prefs by lazy {
        NSUserDefaults.standardUserDefaults().apply {
            persistentDomainForName(DOMAIN_NAME)
        }
    }

    actual fun getString(key: String): String? = prefs.stringForKey(key)

    actual fun setString(key: String, value: String) {
        prefs.setObject(value, forKey = key)
        prefs.synchronize()
    }

    actual fun getInt(key: String, default: Int): Int? = prefs.integerForKey(key).toInt()

    actual fun setInt(key: String, value: Int) {
        prefs.setInteger(value.toLong(), forKey = key)
        prefs.synchronize()
    }

    actual fun getBoolean(key: String, default: Boolean): Boolean? {
        return try {
            prefs.boolForKey(key)
        } catch (e: Exception) {
            default
        }
    }

    actual fun setBoolean(key: String, value: Boolean) {
        prefs.setBool(value, forKey = key)
        prefs.synchronize()
    }

    actual fun remove(key: String) {
        prefs.removeObjectForKey(key)
        prefs.synchronize()
    }

    actual fun clear() {

        prefs.removePersistentDomainForName(DOMAIN_NAME)
        prefs.synchronize()
    }

    actual fun getDouble(key: String, default: Double): Double {

        return try {
            prefs.doubleForKey(key)
        } catch (e: Exception) {
            default
        }
    }

    actual fun setDouble(key: String, value: Double) {
        prefs.setDouble(value, forKey = key)
        prefs.synchronize()
    }

    actual fun getStringList(key: String): List<String>? {
        val array = prefs.arrayForKey(key) as? List<String>
        return array ?: emptyList()
    }

    actual fun setStringList(key: String, values: List<String>) {
        prefs.setObject(values, forKey = key)
        prefs.synchronize()
    }

    actual fun contains(key: String): Boolean = prefs.objectForKey(key) != null
}
